### PR TITLE
Fixed environment variable pass-thru

### PR DESCRIPTION
**Pull request by Haplois**
Fixed environment variable pass-thru from MSBuild task to TestPlatform.

Related https://github.com/microsoft/vstest/issues/669

### CODE REVIEW

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+using Xunit.Abstractions;
+using Microsoft.NET.TestFramework.Utilities;
+using System.Collections.Generic;
+
+namespace Microsoft.DotNet.Cli.Test.Tests
+{
+    public class GivenDotnetTestContainsEnvironmentVariables : SdkTest
+    {
+        private const string TestAppName = "VSTestEnvironmentVariables";
+        private const string EnvironmentOption = "--environment";
```
Comment (wli3) [inline src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:29]: @KathleenDollard need your opinion on a parameter for passing EnvironmentVariables. is "--environment" too ambiguous?

File: src/Cli/dotnet/commands/dotnet-test/Program.cs [hunk 2]
```diff
@@ -167,5 +169,28 @@ private static void UpdateRunSettingsArgumentsText()
             DefaultHelpViewText.Synopsis.AdditionalArguments = " [[--] <RunSettings arguments>...]]";
             DefaultHelpViewText.AdditionalArgumentsSection = LocalizableStrings.RunSettingsArgumentsDescription;
         }
+
+        private static void SetCustomEnvironmentVariables(TestCommand testCommand, AppliedOptionSet optionSet, string optionName)
```
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-test/Program.cs:173]: why optionName need to pass in, this method can only deal with --environment

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs (hunk 1 above)
Comment (Haplois) [reply src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:29]: So just to clarify... That option only applies to `dotnet test` command, and is passed-thru to `vstest.console.exe` which uses the same parameter.

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        private const string EnvironmentVariable1 = "__DOTNET_TEST_ENVIRONMENT_VARIABLE_EMPTY";
+        private const string EnvironmentVariable2 = "__DOTNET_TEST_ENVIRONMENT_VARIABLE_1=VALUE1";
+        private const string EnvironmentVariable3 = "__DOTNET_TEST_ENVIRONMENT_VARIABLE_2=VALUE WITH SPACE";
```
Comment (wli3) [inline src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:32]: @KathleenDollard the syntax is a bit odd. _dotnet test --environment  key1=value1 --environment  key2=value2_

But we don't have something like this before.

**Review by wli3 — COMMENTED** [review]
Some discussion in need in UI. The code looks good. But i am not familiar with dotnet-test

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs (hunk 3 above)
Comment (nohwnd) [reply src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:23]: This is to allow setting env variable just for the target process in unified way. On linux you can already do it this way, but not on windows: 

```
key1=value1 key2=value2 dotnet test
```

So we just take that syntax and use -e or --environment to do the same thing cross platform. Here is my reasoning about the UI https://github.com/microsoft/vstest/issues/669#issuecomment-578813697, and here is the reasoning behind the feature https://github.com/microsoft/vstest/issues/669#issuecomment-578773872. 

We already use a similar syntax for inline settings where you can do `dotnet test -- option1=value1 parent.option2=value2` this get's translated to nodes in the runsettings.xml.

File: src/Assets/TestProjects/VSTestEnvironmentVariables/Tests.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                var (key, value) = ((string)env.Key, (string)env.Value);
+
+                if (!key.StartsWith("__DOTNET_TEST_ENVIRONMENT_VARIABLE_"))
+                    continue;
+
+                Console.WriteLine($"{key}={value}");
+            }
+
+            AssertEnvironmentVariable("__DOTNET_TEST_ENVIRONMENT_VARIABLE_EMPTY", string.Empty);
```
Comment (nohwnd) [inline src/Assets/TestProjects/VSTestEnvironmentVariables/Tests.cs:26]: I would add comment describing how the values are supposed to get here.

File: src/Cli/dotnet/commands/dotnet-test/Program.cs [hunk 5]
```diff
@@ -74,6 +73,9 @@ public static TestCommand FromArgs(string[] args, string[] settings, string msbu
                 noRestore,
                 msbuildPath);
 
+            // Apply environment variables if set
```
Comment (nohwnd) [inline src/Cli/dotnet/commands/dotnet-test/Program.cs:76]:
```suggestion
            // Apply environment variables provided by the user via --environment (-e) parameter, if present
```

File: src/Cli/dotnet/commands/dotnet-test/Program.cs [hunk 6]
```diff
@@ -167,5 +169,30 @@ private static void UpdateRunSettingsArgumentsText()
             DefaultHelpViewText.Synopsis.AdditionalArguments = " [[--] <RunSettings arguments>...]]";
             DefaultHelpViewText.AdditionalArgumentsSection = LocalizableStrings.RunSettingsArgumentsDescription;
         }
+
+        private static void SetCustomEnvironmentVariables(TestCommand testCommand, AppliedOptionSet optionSet)
```
Comment (nohwnd) [inline src/Cli/dotnet/commands/dotnet-test/Program.cs:173]: SetEnvironmentVariablesFromParameters maybe?

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+
+        private readonly string[] EnvironmentVariables = new[] {
+            EnvironmentOption, EnvironmentVariable1,
+            EnvironmentOption, EnvironmentVariable2,
+            EnvironmentOption, EnvironmentVariable3,
+        };
+
+        [Fact]
+        public void TestsShouldBuildAndIncludeEnvironmentVariables()
```
Comment (nohwnd) [inline src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:39]:
```suggestion
        public void ItPassesEnvironmentVariablesFromCommandLineParametersWhenRunningViaCsproj()
```

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                    .Should().Contain("Total tests: 1")
+                    .And.Contain("Passed: 1")
+                    .And.Contain("Passed TestEnvironmentVariables");
+            }
+
+            result.ExitCode.Should().Be(0);
+        }
+
+        [Fact]
+        public void TestsShouldIncludeEnvironmentVariablesDll()
```
Comment (nohwnd) [inline src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:68]:
```suggestion
        public void ItPassesEnvironmentVariablesFromCommandLineParametersWhenRunningViaDll()
```

**Comment by wli3** [comment]
@nohwnd signed off the PR

**Pull request merged**